PXB-1961 Prepare of full backup fails for encrypted tables with Gener…

rahulmalik87 opened the pull request:
…ic error

https://jira.percona.com/browse/PXB-1961

Problem:
PXB is not handling PS encryption types.

Analysis:
PS support addition encrytion types like keyring which is not handled.
This bugs only adresses encryption='n'

Fix:
Handle and process ENCRYPTION_KEY_MAGIC_PS_V3 in PXB

satya-bodapati commented on storage/innobase/include/os0enc.h:
```diff
@@ -83,6 +83,14 @@ class Encryption {
+#ifdef XTRABACKUP
+  static constexpr char KEY_MAGIC_PS_V3[] = "PSC";
+  /* To only  process crypt redo log generated by PS with encryption='n'*/
+  static constexpr uint KEYRING_INFO_MAX_SIZE = 102;
```
I think this is 98 on PS better to use that. Maybe the entire length calculation as is

satya-bodapati commented on storage/innobase/include/os0enc.h:
```diff
@@ -83,6 +83,14 @@ class Encryption {
 
+#ifdef XTRABACKUP
+  static constexpr char KEY_MAGIC_PS_V3[] = "PSC";
+  /* To only  process crypt redo log generated by PS with encryption='n'*/
```
```suggestion
  /* KEYRING ENCRYPTION header size. aka crypt_data. Also present in ENCRYPTION=N Tablespaces*/
```

satya-bodapati commented on storage/innobase/include/os0enc.h:
```diff
@@ -83,6 +83,14 @@ class Encryption {
+  static constexpr char KEY_MAGIC_PS_V3[] = "PSC";
+  /* To only  process crypt redo log generated by PS with encryption='n'*/
+  static constexpr uint KEYRING_INFO_MAX_SIZE = 102;
+  /*to know if encryption is kerying or 'n' */
```
```suggestion
  /* CRYPT_DATA in ENCRYPTION='N" tablespaces always have unencrypted scheme */
```

satya-bodapati commented on storage/innobase/include/os0enc.h:
```diff
@@ -83,6 +83,14 @@ class Encryption {
+  /* To only  process crypt redo log generated by PS with encryption='n'*/
+  static constexpr uint KEYRING_INFO_MAX_SIZE = 102;
+  /*to know if encryption is kerying or 'n' */
+  static constexpr uint CRYPT_SCHEME_UNENCRYPTED = 0;
```
const is fine

altmannmarcelo replied:
Can you please point me to where this is defined on PS?

rahulmalik87 replied:
https://github.com/percona/percona-server/blob/afc5f453aff448993fd5780fb7e9fb281b441978/storage/innobase/include/fil0rkinfo.h#L4

rahulmalik87 commented:
https://pxb.cd.percona.com/view/PXB%208.0/job/percona-xtrabackup-8.0-param-medium/74/ jenkin results. 
will update centos results once dependencies  are resolved.

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1747,10 +1747,34 @@ static byte *recv_parse_or_apply_log_rec_body(
-          if (fil_tablespace_redo_encryption(ptr, end_ptr, space_id,
-                                             start_lsn) == nullptr)
+
+#ifdef XTRABACKUP
```
do we need #ifdef?

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1747,10 +1747,34 @@ static byte *recv_parse_or_apply_log_rec_body(
-                                             start_lsn) == nullptr)
+
+#ifdef XTRABACKUP
+
+          byte *ptr_copy = ptr;
+          ptr_copy += 2;  // skip offset
+          ulint len = mach_read_from_2(ptr_copy);
+          ptr_copy += 2;
+          if (end_ptr < ptr_copy + len) return nullptr;
+
+          /* to only support encryption ='n' with PS */
+          if (memcmp(ptr_copy, Encryption::KEY_MAGIC_PS_V3,
+                     Encryption::MAGIC_SIZE) == 0) {
+            ptr_copy += Encryption::MAGIC_SIZE;
+            uint type = mach_read_from_1(ptr_copy);
+            if (type != Encryption::CRYPT_SCHEME_UNENCRYPTED) {
+              ib::error(ER_IB_MSG_716)
+                  << "Can't take backup of tablespace encrypted with KEYRING "
+                     "encryption";
+              exit(EXIT_FAILURE);
+            }
+
+            ptr += Encryption::KEYRING_INFO_MAX_SIZE;
+            return (ptr);
+          } else if (fil_tablespace_redo_encryption(ptr, end_ptr, space_id,
```
I think problem exists on 5.7 too but encryption header size, magic are different in 5.7 Can you create PR for 5.7 too?

satya-bodapati commented on storage/innobase/xtrabackup/test/t/pxb-1961.sh:
```diff
@@ -0,0 +1,65 @@
+
+vlog "case#1 restore table created with encryption='n'"
+mysql test <<EOF
+set global innodb_page_cleaner_disabled_debug=on;
```
Dont we need have_debug_server.inc?
Also what about disabling checkpoint?

rahulmalik87 replied:
that is present in common.sh

satya-bodapati replied:
why did we put in common.sh? I think we need a separate .sh/.inc for have_debug_server

rahulmalik87 replied:
in pxb all the methods are defined in common.sh like require_xtradb, require_server_version_higher_than, require_debug_pxb_version.

satya-bodapati replied:
the problem then is if a test that needs xtradb, will unnecessarily force them to be debug? :thinking:

rahulmalik87 replied:
these are just functions. It works differently from mtr.

+function is_debug_server()
+{
+
+[[ $MYSQL_VERSION =~ .*debug ]]
+
+}
+

satya-bodapati replied:
ah, we source them. Then you are supposed to call is_debug_server() in the test? and skip it?

rahulmalik87 replied:
right. that's why I have if ! is_debug_server  ; then
        skip_test "Require debug server version"
 fi
 btw Can you review 5.7 first. I have implemented the changes in 5.7 PR. sorry for confusion.

satya-bodapati replied:
I see. Did the review for 5.7

altmannmarcelo commented on storage/innobase/xtrabackup/test/t/pxb-1961.sh:
```diff
@@ -0,0 +1,65 @@
+
+vlog "case#1 restore table created with encryption='n'"
+mysql test <<EOF
+set global innodb_page_cleaner_disabled_debug=on;
```
```suggestion
SET GLOBAL innodb_page_cleaner_disabled_debug=ON;
```

altmannmarcelo commented on storage/innobase/xtrabackup/test/t/pxb-1961.sh:
```diff
@@ -0,0 +1,65 @@
+ALTER TABLE t1 ENCRYPTION='N';
+INSERT INTO t1 VALUES(100);
+EOF
+innodb_wait_for_flush_all
```
Will this ever reach to 0 if we have `set global innodb_page_cleaner_disabled_debug=on;` ?

rahulmalik87 replied:
once we restart server this options gets disabled.

altmannmarcelo commented on storage/innobase/xtrabackup/test/t/pxb-1961.sh:
```diff
@@ -0,0 +1,65 @@
+require_xtradb
+vlog "case#3 try to restore encryption='keyring'"
+mysql test <<EOF
+set global innodb_page_cleaner_disabled_debug=on;
```
Is this necessary since we already do it on case 1?

rahulmalik87 replied:
we need to do it again once the server is restarted

altmannmarcelo commented on storage/innobase/xtrabackup/test/t/pxb-1961.sh:
```diff
@@ -0,0 +1,65 @@
+rm -rf $topdir/backup
+mkdir -p $topdir/backup
+# Backup
+xtrabackup --datadir=$mysql_datadir --backup --target-dir=$topdir/backup 
+# Prepare
+xtrabackup --datadir=$mysql_datadir --prepare --target-dir=$topdir/backup
```
Is `datadir` parameter required here ?

altmannmarcelo commented on storage/innobase/xtrabackup/test/t/pxb-1961.sh:
```diff
@@ -0,0 +1,65 @@
+xtrabackup --datadir=$mysql_datadir --prepare --target-dir=$topdir/backup
+stop_server
+# Restore
+rm -rf $mysql_datadir/ibdata1 $mysql_datadir/ib_logfile* \
```
any reason for selective delete here rather than remove the entire `$mysql_datadir` ?

altmannmarcelo commented on storage/innobase/xtrabackup/test/t/pxb-1961.sh:
```diff
@@ -0,0 +1,65 @@
+verify_db_state test
+
+}
+if ! is_debug_server  ; then
+       skip_test "Require debug server version"
+fi
+
+. inc/common.sh
```
Don't we need `. inc/common.sh` before the  `if ! is_debug_server` for `is_debug_server`  function to be available?

altmannmarcelo commented on storage/innobase/xtrabackup/test/t/pxb-1961.sh:
```diff
@@ -0,0 +1,65 @@
+. inc/common.sh
+
+MYSQLD_EXTRA_MY_CNF_OPTS="
+innodb_file_per_table
```
isn't this default on 8.0?

altmannmarcelo commented on storage/innobase/xtrabackup/test/t/pxb-1961.sh:
```diff
@@ -0,0 +1,65 @@
+
+rm -rf $topdir/backup
+mkdir -p $topdir/backup
+run_cmd_expect_failure $XB_BIN $XB_ARGS --datadir=$mysql_datadir --backup --target-dir=$topdir/backup
```
same as above regarding `datadir`

Commit:
PXB-1961 Prepare of full backup fails for encrypted tables with Generic error

https://jira.percona.com/browse/PXB-1961

Problem:
PXB is not handling redo type generated by PS for encryption keyring

Analysis:
PS supports addition encryption types like keyring which is not handled
by PXB. This bugs only adresses encryption='n'

Fix:
Process KEY_MAGIC_PS generated by PS. also error out if redo log on page
0 found with CRYPT_SCHEME_1 that is keyring encryption is used

rahulmalik87 commented:
https://pxb.cd.percona.com/view/PXB%208.0/job/percona-xtrabackup-8.0-param-medium/82/
the two failing test bug1343722.sh and bug759701.sh doesn't look related to it.
bug1343722.sh appeared in https://pxb.cd.percona.com/view/PXB%208.0/job/percona-xtrabackup-8.0-param-medium/BUILD_TYPE=release,Host=min-centos-8-x64,xtrabackuptarget=xtradb80/68/

satya-bodapati commented on storage/innobase/os/os0enc.cc:
```diff
@@ -198,6 +198,9 @@ void deinit_keyring_services(SERVICE_TYPE(registry) *) { return; }
 constexpr char Encryption::KEY_MAGIC_V1[];
 constexpr char Encryption::KEY_MAGIC_V2[];
 constexpr char Encryption::KEY_MAGIC_V3[];
+#ifdef XTRABACKUP
```
can you get rid of #ifdefs in this entire patch

satya-bodapati commented on storage/innobase/include/os0enc.h:
```diff
@@ -31,6 +31,7 @@ Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA
 #define os0enc_h
 
 #include <mysql/components/my_service.h>
+#include "my_aes.h"
```
why do we need this header?

rahulmalik87 replied:
> why do we need this header?

because of KEYRING_VALIDATION_TAG_SIZE = MY_AES_BLOCK_SIZE;

satya-bodapati replied:
I see

satya-bodapati commented on storage/innobase/include/os0enc.h:
```diff
@@ -98,8 +99,31 @@ class Encryption {
+  /* CRYPT_DATA in ENCRYPTION='N" tablespaces always have unencrypted scheme */
+  static const uint CRYPT_SCHEME_UNENCRYPTED = 0;
+  /* KEYRING ENCRYPTION header size. aka crypt_data. Also present in
+    ENCRYPTION=N Tablespaces*/
```
alignment of ENCRYPTION=N
leave space after Tablespaces

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1852,7 +1868,39 @@ recv_parse_or_apply_log_rec_body(
+				return NULL;
+			}
+
+                        /* With Percona Server, Tables created with
```
comment starts at wrong offset (column)

satya-bodapati replied:
ideally clang-format should have moved this. Can you run clang-format on this file?

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1852,7 +1868,39 @@ recv_parse_or_apply_log_rec_body(
 		if (page_no == 0 && !is_system_tablespace(space_id)) {
-			if (fil_write_encryption_parse(ptr, end_ptr,
+			byte* ptr_copy = ptr;
+			//increase by offset
```
// increase
may be write // skip offset

satya-bodapati commented on storage/innobase/xtrabackup/test/t/keyring_encryption.sh:
```diff
@@ -0,0 +1,48 @@
+. inc/common.sh
+. inc/keyring_file.sh
+
+if ! is_debug_server  ; then
+       skip_test "Require debug server version"
+fi
```
better to put this as function in common.sh as  have_debug() function 
and call that function here.

satya-bodapati replied:
I guess you can address these comments on 5.7 branch and merge to 8.0

satya-bodapati commented on storage/innobase/xtrabackup/test/t/keyring_encryption.sh:
```diff
@@ -0,0 +1,48 @@
+       skip_test "Require debug server version"
+fi
+
+if is_server_version_lower_than 5.7.0
+then
+    skip_test "Requires server version 5.7"
+fi
```
similar to above, have_5_7()

satya-bodapati commented on storage/innobase/xtrabackup/test/t/keyring_encryption.sh:
```diff
@@ -0,0 +1,48 @@
+
+vlog "case#1 restore table created with encryption='n'"
+mysql test <<EOF
+SET GLOBAL innodb_page_cleaner_disabled_debug=ON;
```
why checkpoint thread is not disabled?

satya-bodapati commented on storage/innobase/xtrabackup/test/t/keyring_encryption.sh:
```diff
@@ -0,0 +1,48 @@
+
+vlog "case#3 try to restore encryption='keyring'"
+mysql test <<EOF
+SET GLOBAL innodb_page_cleaner_disabled_debug=on;
```
checkpoint shoud be disabled here too..

satya-bodapati commented on storage/innobase/include/os0file.h:
```diff
@@ -343,6 +343,10 @@ static const char ENCRYPTION_KEY_MAGIC_V2[] = "lCB";
 information version. */
 static const char ENCRYPTION_KEY_MAGIC_V3[] = "lCC";
 
+/* CRYPT_DATA in ENCRYPTION='N" tablespaces always have unencrypted scheme */
```
use single quote around N -> 'N'

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1747,8 +1747,28 @@ static byte *recv_parse_or_apply_log_rec_body(
+            if (type != Encryption::CRYPT_SCHEME_UNENCRYPTED) {
+              ib::error(ER_IB_MSG_716)
+                  << "Can't take backup of tablespace encrypted with KEYRING "
+                     "encryption";
```
better to add << in the next line before encryption.

rahulmalik87 commented:
https://pxb.cd.percona.com/view/PXB%208.0/job/percona-xtrabackup-8.0-multijob/35/

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1747,8 +1747,30 @@ static byte *recv_parse_or_apply_log_rec_body(
+          ptr_copy += 2;
+          if (end_ptr < ptr_copy + len) return nullptr;
+
+          /* With Percona Server, Tables created with
+                        ENCRYPTION='N' have crypt_data
+             (CRYPT_SCHEME_UNENCRYPTED) in Page 0 */
```
possible to align the comment? or is it clang-format messing this up?

rahulmalik87 replied:
it is clang-format

Commit:
Merge branch 'PXB-2.4-1961' into PXB-8.0-1961
